component/loki/source/windowsevent/bookmark.go

```diff
@@ -0,0 +1,83 @@
+//go:build windows
+// +build windows
+
+// This code is copied from Promtail with minor changes.
+
+package windowsevent
+
+import (
+	"bytes"
+	"errors"
+	"github.com/natefinch/atomic"
+	"io"
+	"io/fs"
+	"os"
+
+	"github.com/grafana/loki/clients/pkg/promtail/targets/windows/win_eventlog"
+)
+
+type bookMark struct {
+	handle win_eventlog.EvtHandle
+	isNew  bool
+	path   string
+	buf    []byte
+}
+
+// newBookMark creates a new windows event bookmark.
+// The bookmark will be saved at the given path. Use save to save the current position for a given event.
+func newBookMark(path string) (*bookMark, error) {
+	// 16kb buffer for rendering bookmark
+	buf := make([]byte, 16<<10)
+
+	_, err := os.Stat(path)
+	// creates a new bookmark file if none exists.
+	if errors.Is(err, fs.ErrNotExist) {
+		_, err := os.Create(path)
+		if err != nil {
+			return nil, err
+		}
+		bm, err := win_eventlog.CreateBookmark("")
+		if err != nil {
+			return nil, err
+		}
+		return &bookMark{
+			handle: bm,
+			path:   path,
+			isNew:  true,
+			buf:    buf,
+		}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	// otherwise open the current one.
+	file, err := os.OpenFile(path, os.O_RDWR, 0666)
+	if err != nil {
+		return nil, err
+	}
+	fileContent, err := io.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	fileString := string(fileContent)
+	// load the current bookmark.
+	bm, err := win_eventlog.CreateBookmark(fileString)
+	if err != nil {
+		return nil, err
+	}
+	return &bookMark{
+		handle: bm,
+		path:   path,
+		isNew:  fileString == "",
+		buf:    buf,
+	}, nil
```

Comment:
In this fragment of code few things can still go wrong and possibly the agent will be crashlooping until someone manually repairs the bookmark file or, more likely, deletes it. 

Should we instead have on a high-level something like this?

```
func newBookmark() {
  if not fileExists {
    createNewOne()
  } else {
    err := useExistingFile()
    if err != nil {
      renameCorruptedFile()
      createNewOne()
    }
  }
}
```

The idea being that if we detect a corrupted file, we attempt to recover.

Reply:
If we error on the create bookmark then we could call it again with a blank string to force a new bookmark.